pyccel/parser/semantic.py

```diff
@@ -238,18 +239,27 @@ def annotate(self, **settings):
 
         self._semantic_done = True
 
-        #check if the ast CodeBlock is a Program to add the Garbage collecting
+        # check if the ast CodeBlock is a Program to call the Garbage collecting
         cls = (Header, EmptyNode, NewLine, Comment, CommentBlock, Module)
-        body = ast.body
-        is_module = all(isinstance(i,cls) for i in body)
-        if not isinstance(body[-1], Return) and not is_module:
-            dealloc = [Deallocate(i) for i in self._allocs]
-            body += dealloc
-            ast = CodeBlock(body)
-            self._allocs = []
+        is_module = all(isinstance(i,cls) for i in ast.body)
+        if not is_module:
+            self._ast = ast = self.garbage_collector(ast)
+        self._allocs = []
 
         return ast
 
+    def garbage_collector(self, expr):
+        """
+        Search in a CodeBlock if no trailing Return Node is present add the needed frees.
+
+        Return the same CodeBlock if a trailing Return found otherwise Return a new CodeBlock with additionals Deallocate Nodes.
```

Comment:
typo `additionals`->`additionals`
grammar:
```
 Return the same CodeBlock if a trailing Return found otherwise Return a new CodeBlock with additionals Deallocate Nodes.
```
->
```
Return the same CodeBlock if a trailing Return is found otherwise Return a new CodeBlock with additional Deallocate Nodes.
```